File: DSAlgorithms/patterns/fast_slow_pointers/4_middle_linked_list.py
class node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next
def find_mid(head):
    slow = head
    fast = head
    while(fast is not None and fast.next is not None):
        slow = slow.next
        fast = fast.next.next
    return slow

def main():
    head = node(1)
    head.next = node(2)
    head.next.next = node(3)
    head.next.next.next = node(4)
    head.next.next.next.next= node(5)
    print("middle node: " +  str(find_mid(head).value))

    head.next.next.next.next.next = node(6)
    print("middle node: " +  str(find_mid(head).value))
    
    head.next.next.next.next.next.next = node(7)
    print("middle node: " + str(find_mid(head).value))

main()

File: DSAlgorithms/patterns/fast_slow_pointers/4_middle_linked_list.js
class node{
    constructor(value, next = null)
    {
        this.value = value;
        this.next = next;
    }
}

const find_mid = (head) =>{
    let slow = head;
    let fast = head;
    while((fast !== null && fast.next !== null))
    {
        slow = slow.next;
        fast = fast.next.next;
    }
    return slow;
}

head = new node(1);
head.next = new node(2);
head.next.next = new node(3);
head.next.next.next = new node(4);
head.next.next.next.next = new node(5);
console.log(`middle node: ${find_mid(head).value}`);

head.next.next.next.next.next = new node(6);
console.log(`middle node: ${find_mid(head).value}`);

head.next.next.next.next.next.next = new node(7);
console.log(`middle node: ${find_mid(head).value}`);

